controllers/core/single_obj_controller.go

```diff
@@ -0,0 +1,61 @@
+package core
+
+import (
+	"context"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+)
+
+// NewSingleObjectEventHandler creates a new event handler for a single object.
+// It exploits a WithWatch client to watch for changes on the object and execute a goroutine to listen for the channel
+// and call the handler function when an event occurs.
+// The function is generic and takes two types T and L, where T is the type of the object to watch and L is the type of
+// the list of T objects to watch. The function also takes the name of the object the handler should subscribe to
+// and the namespace to watch.
+// handler is a function that takes the event type and the object that was changed. Event types are defined in watch.go
+// and can be Added, Modified, Deleted, Bookmark and Error (not handled by handler).
+// errorHandler is an optional (nullable pointer to a) function executed when the event type is Error.
+func NewSingleObjectEventHandler[T metav1.Object, L client.ObjectList](ctx context.Context, name string, namespace string, list L,
+	handler func(watch.EventType, T), errorHandler *func(*metav1.Status)) {
+	// TODO: Is there a way for list L to be created from L here with go generics?
+	// TODO: Let's make this crash the container for now when any errors occur here, but leave a TODO to better handle and discuss of it
+
+	cfg := config.GetConfigOrDie()
+
+	cli, err := client.NewWithWatch(cfg, client.Options{})
+	if err != nil {
+		panic(err)
```

Comment:
yeah, we don't want this...maybe just error out

Reply:
Let me change this tomorrow. It was initially meant for the goroutine. But the goroutine is internal to the function now. If the internal goroutine panics, the controller crashes and restarts, which is "good enough" for now and better than having a goroutine dead and the controller still up.

Reply:
is there a way to not panic but send and error and let the pod go through?

Reply:
Removed the panic from the method and returned an error to delegate the user to decide what to do.